src/main/geometries/Tube.java

```diff
@@ -60,9 +73,70 @@ public NormalizedVector normal(Point p) {
 		return sourceToP.subtract(direction().scale(dotProduct)).normalized();
 	}
 
-	@Override
-	public List<Point> intersect(Ray ray) {
-		// TODO Auto-generated method stub
-		return null;
+
+
+	/**
+	 * This function will find intersection points (possibly none) between a {@link Ray} and an {@link Cylinder}.
+	 *
+	 * @param r The {@link Ray} to intersect
+	 * @return a list (possibly empty) of intersection points
+	 */
+	// @Override
+	public List<Point> intersect(Ray r) {
+		Point source;
+		boolean tubeIsCentered = false;
+		if (!axis.source.equals(Point.ORIGIN)) {
+			source = r.source.add(toOrigin);
+		} else {
+			tubeIsCentered = true;
+			source = r.source;
+		}
+		CartesianProductSelf a = new CartesianProductSelf(axis.direction);
+		CartesianProductSelf p = new CartesianProductSelf(source); // p for point
+		CartesianProductSelf v = new CartesianProductSelf(r.direction); // v for vector
+		CartesianProduct pv = new CartesianProduct(source, r.direction); // pv for...
+		// @formatter:off
+		double A =
+			(a.yy * v.xx) - 2*(a.yx * v.yx) + (a.xx * v.yy) +
+			(a.zz * v.yy) - 2*(a.zy * v.zy) + (a.yy * v.zz) +
+			(a.xx * v.zz) - 2*(a.xz * v.xz) + (a.zz * v.xx);
+		double B = 2 * (
+			(a.yy * pv.xx) - a.yx*(pv.xy + pv.yx) + (a.xx * pv.yy) +
+			(a.zz * pv.yy) - a.zy*(pv.yz + pv.zy) + (a.yy * pv.zz) +
+			(a.xx * pv.zz) - a.xz*(pv.xz + pv.zx) + (a.zz * pv.xx));
+		double C = 
+			(a.yy * p.xx) - 2*(a.yx * p.yx) + (a.xx * p.yy) +
+			(a.zz * p.yy) - 2*(a.zy * p.zy) + (a.yy * p.zz) +
+			(a.xx * p.zz) - 2*(a.xz * p.xz) + (a.zz * p.xx) -
+			(radius * radius);
+		// @formatter:on
+		// use abes quadratic thingy instead
+		double det = (B * B) - (4 * A * C);
+		List<Point> intersections = new ArrayList<Point>();
+		if (DoubleCompare.gt(det, 0)) {
+			double t1 = ((-1) * B + Math.sqrt(det)) / (2 * A); // what to do if A is 0?
+			double t2 = ((-1) * B - Math.sqrt(det)) / (2 * A);
+			Ray shiftedRay = new Ray(source, r.direction);
+			fillList(intersections, shiftedRay, tubeIsCentered, fromOrigin, t1);
+			fillList(intersections, shiftedRay, tubeIsCentered, fromOrigin, t2);
+		}
+		if (intersections.isEmpty()) {
+			return Collections.emptyList();
+		}
```

Comment:
it defeats the purpose of using Collections.emptyList() if you're gonna create an empty list enyways. u can just put this in the else
```suggestion
		if (DoubleCompare.gt(det, 0)) {
			List<Point> intersections = new ArrayList<Point>();
			double t1 = ((-1) * B + Math.sqrt(det)) / (2 * A); // what to do if A is 0?
			double t2 = ((-1) * B - Math.sqrt(det)) / (2 * A);
			Ray shiftedRay = new Ray(source, r.direction);
			fillList(intersections, shiftedRay, tubeIsCentered, fromOrigin, t1);
			fillList(intersections, shiftedRay, tubeIsCentered, fromOrigin, t2);
		} else {
			return Collections.emptyList();
		}
```